include/ppx/knob.h

```diff
@@ -0,0 +1,389 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef ppx_knob_h
+#define ppx_knob_h
+
+#include "ppx/command_line_parser.h"
+#include "ppx/config.h"
+#include "ppx/imgui_impl.h"
+#include "ppx/log.h"
+
+#include <iostream>
+#include <unordered_set>
+#include <vector>
+
+namespace ppx {
+
+// ---------------------------------------------------------------------------------------------
+// Knob Classes
+// ---------------------------------------------------------------------------------------------
+
+// Knob is an abstract class which contains common features for all knobs and knob hierarchy
+class Knob
+{
+public:
+    Knob(const std::string& flagName)
+        : mFlagName(flagName), mDisplayName(flagName), mIndent(0) {}
+    virtual ~Knob() = default;
+
+    std::string GetDisplayName() const { return mDisplayName; }
+    std::string GetFlagName() const { return mFlagName; }
+    std::string GetFlagHelp() const { return mFlagDesc; }
+    size_t      GetIndent() const { return mIndent; }
+
+    // Populate flag members
+    void SetDisplayName(std::string displayName) { mDisplayName = displayName; }
+    void SetFlagDesc(std::string flagDesc) { mFlagDesc = flagDesc; }
+    void SetIndent(size_t indent) { mIndent = indent; }
+
+    virtual void Draw() = 0;
+
+    // Flag description in usage message
+    virtual std::string GetFlagHelpText() const = 0;
+
+    // Updates knob value from commandline flag
+    virtual void UpdateFromFlag(const CliOptions& opts) = 0;
+
+    // Default values are the values when the application starts. The knobs can be reset to default by a button in the UI
+    virtual void ResetToDefault() = 0;
+
+protected:
+    std::string mDisplayName;
+    std::string mFlagName;
+    std::string mFlagDesc;
+    size_t      mIndent;
+};
+
+// KnobCheckbox will be displayed as a checkbox in the UI
+// Currently supports variants of type: bool
+template <typename T, std::enable_if_t<std::is_same_v<T, bool>, bool> = true>
+class KnobCheckbox final
+    : public Knob
+{
+public:
+    KnobCheckbox(const std::string& flagName, T defaultValue)
+        : Knob(flagName)
+    {
+        SetDefault(defaultValue);
+    }
+
+    void Draw() override
+    {
+        ImGui::Checkbox(mDisplayName.c_str(), &mValue);
+    }
+
+    void ResetToDefault() override { SetValue(mDefaultValue); }
+
+    // GetFlagHelpText format:
+    // --flag_name <true|false> : flag description
+    std::string GetFlagHelpText() const override
+    {
+        std::string flagHelpText = "--" + mFlagName + " <true|false>";
+        if (!mFlagDesc.empty()) {
+            flagHelpText += " : " + mFlagDesc;
+        }
+        return flagHelpText + "\n";
+    }
+
+    // Updates mValue and mDefaultValue, expected flag format:
+    // --flag_name <true|false>
+    void UpdateFromFlag(const CliOptions& opts) override
```

Comment:
nit: UpdateFromFlags? or rename to UpdateFromFlag in the manager? (plural is inconsitent)